### PR TITLE
Provide correct warp size to compiler

### PR DESCRIPTION
**Pull request by joviliast**
Make warp size parameter and kernel launching dependent on architecture

Signed-off-by: joviliast <iveselov.nn@gmail.com>

### CODE REVIEW

File: python/triton/third_party/hip/hip_backend.py [hunk 1]
```diff
@@ -274,22 +274,29 @@ def get_amdgcn_bitcode_paths(gfx_arch: str):
...
+def get_arch_name() -> str:
+    arch_info = _triton.get_arch_info()
+    gfx_arch_details = re.search('amd.*', arch_info).group(0).strip().split('--')
+    arch_name_features = gfx_arch_details[1].split(':')
+    return arch_name_features[0]
+
+
+def get_warp_size(gpu_name: str) -> int:
+    # 64 is not supported for RDNA for now
+    if 'gfx1' in gpu_name:
```
Comment (zhanglx13) [inline python/triton/third_party/hip/hip_backend.py:286]: It's better to use the full gfx generation name, like gfx10, gfx9

File: python/triton/third_party/hip/hip_backend.py [hunk 2]
```diff
@@ -128,7 +128,7 @@ def format_of(ty):
   // printf("_launch hip kernel\\n");
   void *params[] = {{ {', '.join(f"&arg{i}" for i in params)} }};
   if (gridX*gridY*gridZ > 0) {{
-      HIP_CHECK(hipModuleLaunchKernel(function, gridX, gridY, gridZ, 64*num_warps, 1, 1, shared_memory, stream, params, 0));
+      HIP_CHECK(hipModuleLaunchKernel(function, gridX, gridY, gridZ, {str(warp_size)}*num_warps, 1, 1, shared_memory, stream, params, 0));
```
Comment (zhanglx13) [inline python/triton/third_party/hip/hip_backend.py:131]: why do you need a string here?

**Review by zhanglx13 — COMMENTED** [review]
I have a general question. How does this PR plan to live with upstream? Do we need to "copy" this PR upstream directly? Or do we need to redo everything?

**Comment by scxiao** [comment]
We have a PR [pull/244/](https://github.com/ROCmSoftwarePlatform/triton/pull/244/files) that did this before. It is a complete solution, maybe you can take a look @joviliast. Thanks

**Comment by joviliast** [comment]
> I have a general question. How does this PR plan to live with upstream? Do we need to "copy" this PR upstream directly? Or do we need to redo everything?

Some of changes are similar, but they can't be copied. I'll create another PR in upstream. Is it Ok?

**Comment by joviliast** [comment]
> We have a PR [pull/244/](https://github.com/ROCmSoftwarePlatform/triton/pull/244/files) that did this before. It is a complete solution, maybe you can take a look @joviliast. Thanks

Thank you. I'll take it as a reference

**Review by scxiao — APPROVED** [review]

**Comment by zstreet87** [comment]
LGTM

**Comment by zhanglx13** [comment]
@zstreet87 We already have a PR upstream dealing with warpSize. Why do we need the same thing (or something very similar) in this fork?

File: python/triton/third_party/hip/hip_backend.py [hunk 3]
```diff
@@ -274,22 +274,30 @@ def get_amdgcn_bitcode_paths(gfx_arch: str):
...
     return amdgcn_bitcode_paths
 
+def get_arch_name() -> str:
+    arch_info = _triton.get_arch_info()
+    gfx_arch_details = re.search('amd.*', arch_info).group(0).strip().split('--')
+    arch_name_features = gfx_arch_details[1].split(':')
+    return arch_name_features[0]
+
+
+def get_warp_size(gpu_name: str) -> int:
```
Comment (xinyazhang) [inline python/triton/third_party/hip/hip_backend.py:284]: You don't need to guess the `warp_size` from the arch info string. Triton has a `get_warp_size` function defined in C side.
https://github.com/ROCm/triton/blob/e9ab8a27197f03b947ff2561f5fcb88d8cc88dbb/python/src/triton.cc#L1997-L2002
https://github.com/ROCm/triton/blob/e9ab8a27197f03b947ff2561f5fcb88d8cc88dbb/include/triton/Tools/Sys/GetPlatform.hpp#L251-L292

File: python/triton/third_party/hip/hip_backend.py [hunk 4]
```diff
@@ -317,11 +324,12 @@ def get_amdgpu_arch_fulldetails():
         gfx_arch = os.environ.get('MI_GPU_ARCH', arch_name)
         if gfx_arch is None:
             raise RuntimeError('gfx_arch is None (not specified)')
-        mat_core_ver = gpu_matrix_core_version()
-        capability = gpu_matrix_core_version() * 100
+        mat_core_ver = gpu_matrix_core_version(arch_name)
```
Comment (xinyazhang) [inline python/triton/third_party/hip/hip_backend.py:327]: Is this change strictly required for configurable warp size? It seems to be an orthogonal change that fixes the behavior of `MI_GPU_ARCH`
(Furthermore I don't think this env var is well named, I would use `TRITON_OVERRIDE_GPU_ARCH` or sth similar)

**Review by xinyazhang — CHANGES_REQUESTED** [review]
Replace `get_warp_size` with `_triton.get_warp_size`.
If you want to enforce `MI_GPU_ARCH` (which should be a separate PR and requires more discussions), add it to the C implementation.

File: python/triton/third_party/hip/hip_backend.py (hunk 3 above)
Comment (scxiao) [reply python/triton/third_party/hip/hip_backend.py:284]: @joviliast, could you please take a look at this comments?

File: python/src/triton.cc [hunk 5]
```diff
@@ -1997,7 +1997,8 @@ void init_triton_translation(py::module &m) {
   m.def(
       "get_warp_size",
       []() {
-        return std::get<1>(getArchInfo());
+        if (userDefArch.empty())
```
Comment (xinyazhang) [inline python/src/triton.cc:2000]: How did you get this compiled? I can't find its definition or initialization.

Also I still recommend making another PR to enable `MI_GPU_ARCH`, because it needs to address quite a few considerations, for example, the name of the variable itself is questionable due to potential naming conflicts.

**Review by xinyazhang — APPROVED** [review]
LGTM once GPU Unit Tests passed (currently it cannot be built)

File: python/src/triton.cc (hunk 5 above)
Comment (joviliast) [reply python/src/triton.cc:2000]: Accidental change of user defined arch. sorry,

I'll create PR to enable user defined arch today or tomorrow
